### PR TITLE
[16.0][MIG] stock_picking_invoicing: Migration to 16.0

### PR DESCRIPTION
**Pull request by mbcosta**
Standard migration of stock_picking_invoicing to v16, https://github.com/OCA/account-invoicing/issues/1250 some changes was necessary:

- To avoid error in the Views with the Group parameter,  I'm just create a simply method to check if the user has the group Account and make field Readonly if not, log of the error:

Field 'invoice_state' used in attrs ({'invisible': [('invoice_state', 'in', [False, 'none'])]}) is restricted to the group(s) account.group_account_invoice.

-  Removed get Price Unit for Invoice methods,  because the field become compute in the account.move.line https://github.com/OCA/OCB/blob/16.0/addons/account/models/account_move_line.py#L330

- Refund cases can has Local Origin Usage as Customer and Internal

- Included onchange methods to the simulation the creation of the Invoice

- In the cases of Reversal Moves is necessary link the account.move.lines with the stock.move related because the field picking_ids in account.move become compute https://github.com/OCA/stock-logistics-workflow/blob/16.0/stock_picking_invoice_link/models/account_move.py#L18

This PR should be replace https://github.com/OCA/account-invoicing/pull/1449 because it's based in the v15 and by the tests I'm made with demo data the module are working as expect.

cc @renatonlima @rvalyi @jguenat

### CODE REVIEW

**Comment by jguenat** [comment]
Thanks @mbcosta  you beat me to it I was planning to work on this next week :)
I'll review this more thoroughly Tuesday but here is my first comment:

"To avoid error in the Views with the Group parameter, I'm just create a simply method to check if the user has the group Account"

I think this is a little bit overkill and it seems that the standard "odoo way" of doing it is to simply add the field a second time with invisible="1". It mentioned in the comments of this PR https://github.com/odoo/odoo/pull/95729

File: stock_picking_invoicing/models/stock_move.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+    @api.model
+    def _get_account(self, fiscal_position, account):
+        """
+        Map the given account with given fiscal position
+        :param fiscal_position: account.fiscal.position recordset
+        :param account: account.account recordset
+        :return: account.account recordset
+        """
+        return fiscal_position.map_account(account)
```
Comment (jguenat) [inline stock_picking_invoicing/models/stock_move.py]: Seems it's not used anymore

**Review by jguenat — COMMENTED** [review]
LG thanks

**Comment by mbcosta** [comment]
Thanks @jguenat , you are right about the new way of use the parameter groups in view, I made the change by removed my commit and included a new one based in your PR with the dependecy of base_view_inheritance_extension module as you did, by this reason I put you as the author of the commit https://github.com/OCA/account-invoicing/pull/1465/commits/45dad5750f86f95ea1eb18cccca5de42f67fc5a6 and made a rebase in PR, I'll check the necessity to keep the _get_taxes and _get_account methods .

**Comment by mbcosta** [comment]
@jguenat the methods _get_taxes and _get_account was removed in the last commit, as my focus are the Brazilian Localization I have tried to avoid changes or remove methods that can be used for other scenarios, in Brazil case it's not being used, if in other scenarios someone see an reason to keep it, should be identify to keep the code.

It's important to say, despite the Price Unit became an Computed Field in account.move.line object https://github.com/odoo/odoo/blob/16.0/addons/account/models/account_move_line.py#L330, there are the possibility to change the value by inherit the method _compute_price_unit https://github.com/odoo/odoo/blob/16.0/addons/account/models/account_move_line.py#L842 and when call _get_tax_included_unit_price method https://github.com/odoo/odoo/blob/16.0/addons/account/models/product.py#L118 send the parameter product_price_unit https://github.com/odoo/odoo/blob/16.0/addons/account/models/product.py#L137 , if necessary I can modify the commit https://github.com/OCA/account-invoicing/pull/1465/commits/4704544eed4fd4283150e652dc12e736fad4d5f8 to keep in this module the possibility to inform and change the Price Unit in account.move.line ( cc @renatonlima @rvalyi )

**Comment by jguenat** [comment]
Thank you @mbcosta 
In previous versions this module had a method to compute invoice line price unit so I think it's best to do as you sugest.
It was also pricelist sensitive so we have a loss in funtionality here. We could simply document it or try to reimplement it.
I can work on your suggestion with pricelist integration in a few days if you think we should proceed in this direction.

**Comment by mbcosta** [comment]
@jguenat I agree to keep this functionality, by this reason I remade the commit https://github.com/OCA/account-invoicing/pull/1465/commits/da33d7544a5a68fa4ffc3f145c645c150787d9f2 adapting the code as described before, tests and suggestions are welcome.

File: stock_picking_invoicing/models/stock_move.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+            if partner and partner.property_product_pricelist:
+                product = product.with_context(
+                    partner=partner.id,
+                    quantity=qty,
+                    pricelist=partner.property_product_pricelist.id,
+                    uom=fields.first(self).product_uom.id,
+                )
+                result = product.lst_price
+            else:
+                result = product.lst_price
```
Comment (jguenat) [inline stock_picking_invoicing/models/stock_move.py]: This always returns product.lst_price.
Getting price from pricelist is rather complicated https://github.com/odoo/odoo/blob/d44106309c4a11fc3c0dca4ed8457e1f27bd1527/addons/sale/models/sale_order_line.py#L427
My idea of re implementing pricelist support in this module seems bad. I suggest using account_invoice_pricelist if the project requires it.

Keeping this method as hook for other modules is a good idea though.

```suggestion
        else:
                result = product.lst_price
```

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+                line_values = False
+                for moves in moves_list:
+                    line_values = self._get_invoice_line_values(
+                        moves, invoice_values, invoice
+                    )
+                    if line_values:
+                        lines.append((0, 0, line_values))
+                if line_values:  # Only create the invoice if it has lines
+                    invoice_values["invoice_line_ids"] = lines
+                    invoice_values["invoice_date"] = self.invoice_date
```
Comment (jguenat) [inline stock_picking_invoicing/wizards/stock_invoice_onshipping.py:528]: Having both line_ids and invoice_line_ids causes a AssertionError when sanitizing the invoice dict.
```suggestion
                    invoice_values["invoice_date"] = self.invoice_date
                    invoice_values["line_ids"] = []
```

**Comment by mbcosta** [comment]
@jguenat I changed the method to Get Price Unit to Invoice to consider the Partner Pricelist and the Seller Price, based in the Sale and Purchase modules, the results and cases are describe in the images below:

OBS.: Just for local tests I put the field Price Unit visible in the Stock View but not in the code here, should we included this field in the view? I think yes, if necessary , to the users be able to change.

- Sale 
Price Unit is 0.0, code consider the value was not informed
image

Invoice
image

Value of Field Price Unit informed by User should has priority
image

Invoice
image
 
Product has Price Unit in the Partner Pricelist
image

Invoice
image

- Purchase
Price Unit 0.0, consider value not informed
image

Invoice
image

Field Price Unit informed by User
image

Invoice
image

Partner in stock.move are in the list of Sellers of this Product ( to be possible see the field seller_ids is necessary install module Purchase, the module Product create the field but only module Purchase include it in the view )
image

But not in this
image

Picking
image

Invoice
image


By the tests it's seems unnecessary include the dependecy of account_invoice_pricelist module, I also made small changes in Demo Data and Tests to check the cases described, but if I missing some case let me know.

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py (hunk 3 above)
Comment (mbcosta) [reply stock_picking_invoicing/wizards/stock_invoice_onshipping.py:528]: Change made as susgest, thanks

**Comment by jguenat** [comment]
Thank you @mbcosta for the work, tested it and it works well with my use case.
You could add the Unit Price in the view with optional="hide". Fine by me.

**Review by jguenat — APPROVED** [review]

**Comment by mbcosta** [comment]
Thanks @jguenat for your review, I included the field Price Unit with optional="hide" and change the name of View Files to standard views/<main_model>_views.xml https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst

**Review by rvalyi — CHANGES_REQUESTED** [review]
Hello @mbcosta thank you for this large work. However, I think this account-invoicing repo is the kind of very heterogeneous repo where heroic maintainers like @pedrobaeza told us several times they find it difficult to deal with so many commits (eventually in some other repos like l10n-brazil or mis-buillder the maintainers are more specialized and it's OK to get more details).

There are 13 commits just for the migration. So I suggest you squash a few of them. I identified these 4 trivial commits that can be squashed into the main migration commit:

https://github.com/OCA/account-invoicing/pull/1465/commits/aa784dd9f9ebda2bf15ae1fa3d2d21aa339353f1
[REF] stock_picking_invoicing: Standard File Name for Views.
  
https://github.com/OCA/account-invoicing/pull/1465/commits/11e2e6cfefceae51e0b1f4818b8796e9b902ab9e
[MIG] stock_picking_invoicing: Remove unused methods.

https://github.com/OCA/account-invoicing/pull/1465/commits/f64cf2cc023f21beec3040360f6e35e8f5b7cd9b
[MIG] stock_picking_invoicing: In Refund cases the Local Origin Usage can be Customer or Internal


https://github.com/OCA/account-invoicing/pull/1465/commits/d4bc3e8aba494ba0c820c1253e8d5945b1670e09
[MIG] stock_picking_invoicing: In object stock.picking field move_lines was altered to move_ids


I think you can merge them into:

https://github.com/OCA/account-invoicing/pull/1465/commits/0e7d884d9766593d32e461ac627748288e36ae0a
[16.0][MIG] stock_picking_invoicing: Migration to 16.0

A way to do it:

1. first save a backup branch for not loosing your work
2. for each of the 4 commits:

- git checkout `<sha1>`
- git reset HEAD~1 --soft
- git stash  # this ill save the diff of the commit
- git checkout 16.0-mig-stock_picking_invoicing  # back to the normal branch to prepare the rebase
- git rebase -i 41109ab172d64f4b849fd8741a1e947037409e2a  # the commit before the migration; then indicate you will edit (e) the migration commit ([MIG] stock_picking_invoicing: Migration to 16.0) and indicate you will drop (d) the commit you already put in the stash
- start the rebase
- git stash pop  # will put the content of the stash in the work  tree to be committed.
- git commit . --amend
- git rebase --continue
- repeat of the 3 other commits.

ping we if you have trouble. cc @renatonlima @marcelsavegnago

**Comment by jguenat** [comment]
> Hello @mbcosta thank you for this large work. However, I think this account-invoicing repo is the kind of very heterogeneous repo where heroic maintainers like @pedrobaeza told us several times they find it difficult to deal with so many commits (eventually in some other repos like l10n-brazil or mis-buillder the maintainers are more specialized and it's OK to get more details).
> 
> There are 13 commit just for the migration. So I suggest you squash a few of them. I identified these 4 trivials commits that can be squashed into the main migration commit:
> 
> [aa784dd](https://github.com/OCA/account-invoicing/commit/aa784dd9f9ebda2bf15ae1fa3d2d21aa339353f1) [REF] stock_picking_invoicing: Standard File Name for Views.
> 
> [11e2e6c](https://github.com/OCA/account-invoicing/commit/11e2e6cfefceae51e0b1f4818b8796e9b902ab9e) [MIG] stock_picking_invoicing: Remove unused methods.
> 
> [f64cf2c](https://github.com/OCA/account-invoicing/commit/f64cf2cc023f21beec3040360f6e35e8f5b7cd9b) [MIG] stock_picking_invoicing: In Refund cases the Local Origin Usage can be Customer or Internal
> 
> [d4bc3e8](https://github.com/OCA/account-invoicing/commit/d4bc3e8aba494ba0c820c1253e8d5945b1670e09) [MIG] stock_picking_invoicing: In object stock.picking field move_lines was altered to move_ids
> 
> I think you can merge them into:
> 
> [0e7d884](https://github.com/OCA/account-invoicing/commit/0e7d884d9766593d32e461ac627748288e36ae0a) [16.0][MIG] stock_picking_invoicing: Migration to 16.0
> 
> A way to do it:
> 
> 1. first save a backup branch for not loosing your work
> 2. for each of the 4 commits:
> 
> * git checkout `<sha1>`
> * git reset HEAD~1 --soft
> * git stash  # this ill save the diff of the commit
> * git rebase -i [41109ab](https://github.com/OCA/account-invoicing/commit/41109ab172d64f4b849fd8741a1e947037409e2a)  # the commit before the migration; then indicate you will edit (e) the migration commit ([MIG] stock_picking_invoicing: Migration to 16.0) and indicate you will drop (d) the commit you already put in the stash
> * start the rebase
> * git stash pop  # will put the content of the stash in the work  tree to be committed.
> * git commit . --amend
> * git rebase --continue
> * repeat of the 3 other commits.
> 
> ping we if you have trouble. cc @renatonlima @marcelsavegnago

Hello @rvalyi ,
I profit of your git expertise to ask as question.
In this case I would have done a git rebase -i "commit before migration"
then rearrange the 4 commits to squash just below the migration commit and change pick to fixup.
Would it have worked?
I tried it there https://github.com/jguenat/account-invoicing/commits/16.0-mig-stock_picking_invoicing-test-rebase
Thank you

**Comment by mbcosta** [comment]
Thanks @rvalyi for your review, I merged the commits you mention, except "[REF] stock_picking_invoicing: Standard File Name for Views." because it's seems not possible to rename and change the files in the same commit, if necessary I can look to merge more commits in the Migration one. By some reason the code coverage reduces, I also will look to keep the PR green with more tests.

**Comment by mbcosta** [comment]
About the last changes:

- It's seems unnecessary inherit account.move.line objetc and _compute_price_unit method as I did before, tests are checking the cases of Price Unit informed by User in Picking or get Price from Sellers or from PriceList, the commit was removed

- Despite there are methods to change Invoice State for Invoiced or Not Billable the buttons at some point in the other versions was removed, I didn't noticed before, so I put it back

image

And include onchage method to update the Invoice State when the field in the Picking are changed

image

- Changed the product used in Demo Data to avoid error of Trancking Lot when MRP module are install, v14 PR https://github.com/OCA/account-invoicing/pull/1547 , but different from PR I choosed use a Product with BoM to, if necessary, test this case.

-Refactoring Tests by create methods to avoid duplicate code and, as said before, included tests for Price Unit cases.

**Comment by mbcosta** [comment]
We should avaliate the necessity of the Buttons "Set Invoiced" and "Set Not Billable" in Picking View. In my opinion the case of Not Billable is valide to allow users "fix" when a Picking was wrong marked "To Be Invoiced", I'm not certain about the case of "Invoiced", in which cases it will be use.

**Review by hildickethan — APPROVED** [review]
code+functional review, but I think it's more standard procedure to squash most of these migration commits unless they are adding new features
a86ecf5100dde9f1e18fb54b8db0c896541a196f, e713989359905793438b541bf9fad8a55245f875, 5ed7932c230519d113040f73dfdb9ca2b2d4d4e0, 473c11978c219541edc37d08c58ed7a983f25499, b0d65e71a9a0200960f98805584656fb32347a9b, 273d4cba334dee290981cbd44871369d8f9877d6, 76f387fd4b9031e2e91790ab75cf51955d5af635 and 76f259bee4a73abfcf62763c6fbd61d1ada50171 could all be squashed as generic migration
6af8bb2c4e06f8145dc00d036220280a9a2631ce and 573b5a2f407186196c3eb52d2f47ba453899a644 i think are fine as is seperately as new features

File: stock_picking_invoicing/wizards/stock_invoice_onshipping.py [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+        return values
+
+    def _get_invoice_line_values(self, moves, invoice_values, invoice):
+        """
+        Create invoice line values from given moves
+        :param moves: stock.move
+        :param invoice: account.move
+        :return: dict
+        """
+        name = ", ".join(moves.mapped("name"))
```
Comment (adrip-s73) [inline stock_picking_invoicing/wizards/stock_invoice_onshipping.py]: hi, @mbcosta we found a problem when invoicing a picking with multiple moves with the same product it duplicates the name of the product in the invoice line and if we have the group = "partner_product" . We think that if changing this line to only get the first move's description if using the partner_prooduct option should avoid the issue.

**Comment by mbcosta** [comment]
Thanks @hildickethan-S73 @adrip-s73 for your review, about the problems:

- Squash 4 commits 

image

commit  https://github.com/OCA/account-invoicing/pull/1465/commits/8c69b40eca60c214ee6a7af5b8c246ce1b0410b1 has a different author

- Duplicate field Name/Label in Invoice Line in the case of Grouped Pickings by Partner/Product, 

image

In the last commit I tried to fix it, I let separated commit to allow a backport, if necessary, or to made some change after reviews.

image

**Review by rvalyi — APPROVED** [review]
kudos for the hard work @mbcosta! ok for me, I think we can merge now as @jguenat approved and as you fixed the issue spotted 2 weeks ago.

**Comment by mbcosta** [comment]
thank all for review, if you find some problem or has a doubt about this module please mark me in the issue or PR, I have worked in this module for a while because it's important for Brazilian Localization so we need to keep it functional, but we also work to keep compatible with another countries.

**Comment by jguenat** [comment]
thank you @mbcosta for your work !

**Pull request merged**